CONTRACTS: Dynamic frame condition checking

remi-delmas-3000 opened the pull request:
# Dependencies
The first 6 commits are from these other PRs (do not read)
- #7082
- #7083
- #7084
- #7085
- #7086
- #7091
- #7214 


The build fails with ubuntu 18.04 and I am not sure why, since no error is visible on the output.

Implement the functionality described below.

# Dynamic frame condition checking (DFCC) for function contracts

## Motivation

This new implementation proposes front-end extensions and a new dynamic method for checking assigns and frees clauses addressing all of the above:

* we allow parametric assigns clauses, frees clauses in the contracts language, 
* we allow function calls in pre and post conditions contracts but check them for side effects when they are used
* the checking new method is dynamic and does not require inlining the program or unwinding loops a priori,
* the checking new method is compatible with deferred resolution of function pointers calls during symex,
* the instrumentation hooks are written as C library,
* code duplication for contract checking and contract replacement has been reduced,
* all CPROVER library functions are supported,
* checking contracts on recursive functions is now supported (by assuming preconditions and asserting post conditions on the top level call and asserting pre conditions and assuming post conditions on the recursive call)
* the user documentation for contracts was moved to the contracts/doc folder (doxygen)
* the developer documentation/architecture documentat is found contracts/doc folder (doxygen)

## Current limitations:
* loop contracts are not yet ported to the new framework
* the deferred function pointer resolution is not yet implemented, user-specified function pointer restriction  is still required
* havocing during contract replacement still relies on pointer snapshots known to affect performance
* more work is performed by symbolic execution which can be felt on runtime
* source locations for the instrumented code are still very partial and messy

## Frame condition checking

Let’s call `f_top`​ the top-level function against which the contract is being checked. 

We assume `f_top`​ is itself called from a `harness` function, which may call other functions besides `f_top`​.

Let’s consider a function `f` that is either `f_top​` or one of the functions called (directly or indirectly) by `f_top`​. The goal is to verify that `f` writes only to memory locations that are part of the frame of the call to `f_top` but allowed by the contract or that were allocated after entering `f_top`​ and are hence not part of the frame of the call. To do that, we add ghost code to the program to build representations the following sets of memory locations as the program runs:

* (`contract_assigns`, `contract_frees`, `allocated`, `deallocated`)
where:
* `contract_assigns` is the set of memory locations specified in the *assigns clause* of the contract being checked
* `contract_frees` is the set of pointers specified in the *frees clause* of the contract being checked
* `allocated` is the set objects identifiers allocated on the stack or the heap since entering `f_top`​
* `deallocated`  is the set of heap objects identifiers deallocated since entering `f_top`​ (allows verification that an object was freed as a post-condition of the contract)

The instrumentation transforms all user-defined goto functions and library functions:

```c
ret_t f(<parameters>);
```

Into functions accepting an extra parameter that is a pointer to a dynamic representation of the sets introduced above:

```c
ret_t f(<parameters>,  __CPROVER_contracts_write_set_ptr_t write_set);
```

Function bodies is instrumented so that 
- when the write_set pointer is NULL no checks are performed,
- when it is not NULL the following checks are performed
  * Assignments LHS are checked for inclusion in `contract_assigns` or `allocated`,
  * DECL objects are recorded in `allocated`
  * Dynamically allocated objects are recorded in `allocated`
  * DEAD objects or objects assigned to `__CPROVER_dead_object` are removed from `allocated`
  * dynamic objects deallocated with `__CPROVER_deallocate` are checked for inclusion in `contract_frees` or `allocated`, and recorded in `deallocated` (so that if needed we can check in post conditions that the function indeed deallocated the expected memory).
  * The current write set is propagated to all functions calls or function pointer calls.

Since the write set is dynamically updated, a same function called at different call sites can see a different write set.

Write sets are initialised from contracts descriptions and injected in function calls that are checked against contracts, or used to perform inclusion checks against write sets obtained from callers for functions that are replaced by a contract. 

* * *

qinheping commented on regression/contracts/assigns-slice-targets/main-enforce.c:
```diff
@@ -7,11 +7,14 @@ struct st
-void foo(struct st *s)
+void foo(struct st *s, struct st *ss)
   // clang-format off
   __CPROVER_requires(__CPROVER_is_fresh(s, sizeof(*s)))
```
To be consistent, we may also want to require __CPROVER_requires(__CPROVER_is_fresh(ss, sizeof(*ss))).

qinheping commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -845,9 +857,15 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
     typecheck_spec_assigns(
       static_cast<unary_exprt &>(code.add(ID_C_spec_assigns)).op().operands());
   }
+
+  if(code.find(ID_C_spec_frees).is_not_nil())
+  {
+    typecheck_spec_frees(
+      static_cast<unary_exprt &>(code.add(ID_C_spec_frees)).op().operands());
+  }
```
We are doing the same checks for loop contracts in different kind of loops. Please consider to wrap all type checks for loop contracts (including `typecheck_spec_loop_invariant(code);
typecheck_spec_decreases(code);`) in a new function `typecheck_loop_contract(codet &)`.

qinheping commented on src/ansi-c/cprover_builtin_headers.h:
```diff
@@ -124,6 +128,9 @@ __CPROVER_bool __CPROVER_overflow_unary_minus();
+  __CPROVER_bool is_ptr_to_ptr);
+__CPROVER_assignable_t __CPROVER_whole_object(void *ptr);
+__CPROVER_assignable_t __CPROVER_object_from(void *ptr);
+__CPROVER_assignable_t __CPROVER_object_upto(void *ptr, __CPROVER_size_t size);
```
❓ Why couldn't `__CPROVER_typed_target` be a builtin function? Is it because it is a polymorphic function?

qinheping commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4276 @@
+  // true iff elems is indexed by the object id of the pointers
+  __CPROVER_bool indexed_by_object_id;
+  // array of void *pointers, indexed by their object ID
+  // or some other order
+  void **elems;
+} __CPROVER_contracts_obj_set_t;
+
+typedef __CPROVER_contracts_obj_set_t *__CPROVER_contracts_obj_set_ptr_t;
+
+typedef struct __CPROVER_contracts_write_set_t
+{
+  // set of car derived from the contract (by rank)
+  __CPROVER_contracts_car_set_t contract_assigns;
+  // set of freeable pointers derived from the contract (by object id)
+  __CPROVER_contracts_obj_set_t contract_frees;
+  // set of freeable pointers derived from the contract (by rank)
+  __CPROVER_contracts_obj_set_t contract_frees_replacement;
+  // set of objects allocated by the function under analysis (by object id)
+  __CPROVER_contracts_obj_set_t allocated;
+  // set of objects deallocated by the function under analysis (by object id)
+  __CPROVER_contracts_obj_set_t deallocated;
+  // object set supporting the is_freshr predicate checks (by object id)
+  __CPROVER_contracts_obj_set_t is_freshr_seen;
+  // object set recording the is_freshr allocations in post conditions
+  // (replacement mode only)
+  __CPROVER_contracts_obj_set_ptr_t linked_allocated;
+  // true iff this write set is used for contract replacement
+  __CPROVER_bool replacement;
+  // true iff this write set checks requires clauses in an assumption ctx
+  __CPROVER_bool assume_requires_ctx;
+  // true iff this write set checks requires clauses in an assertion ctx
+  __CPROVER_bool assert_requires_ctx;
+  // true iff this write set checks ensures clauses in an assumption ctx
+  __CPROVER_bool assume_ensures_ctx;
+  // true iff this write set checks ensures clauses in an assertion ctx
+  __CPROVER_bool assert_ensures_ctx;
+} __CPROVER_contracts_write_set_t;
+
+typedef __CPROVER_contracts_write_set_t *__CPROVER_contracts_write_set_ptr_t;
+#endif
```
Do we have to repeat this code snippet 43 times? It make this file hard to read.

jimgrundy replied:
+1 reading this is kind of brutal

qinheping commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4276 @@
+  __CPROVER_size_t idx = set->max_elems;
+  __CPROVER_contracts_car_t *elem = set->elems;
+CAR_SET_CONTAINS_LOOP:
+  while(idx != 0)
```
Stronger guard `while(idx != 0 || incl)` for early break.

jimgrundy replied:
This code is only for CBMC analysis, not for running, and early early makes analysis slower.

qinheping commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4276 @@
+  __CPROVER_size_t idx = set->contract_assigns.max_elems;
+  __CPROVER_contracts_car_t *elem = set->contract_assigns.elems;
+SET_CHECK_ASSIGNMENT_LOOP:
+  while(idx != 0)
```
`while(idx != 0 || incl)`

jimgrundy replied:
Perhaps unexpectedly, that will slow analysis down.

qinheping commented on src/goto-instrument/contracts/cfg_info.h:
```diff
@@ -194,4 +194,48 @@ class loop_cfg_infot : public cfg_infot
+    dirty.insert(dirty_ids.begin(), dirty_ids.end());
+  }
+
+  /// Returns true iff `ident` is a loop local.
```
Should it be goto-program local?

qinheping commented on src/goto-instrument/contracts/dfcc.cpp:
```diff
@@ -0,0 +1,457 @@
+  const std::string &harness_id,
+  const std::set<std::string> &to_check,
+  const std::set<std::string> &to_replace,
+  const bool apply_loop_contracts,
```
What is the purpose of `apply_loop_contracts`? I didn't see any use of it in these functions. Did I miss something?

jimgrundy commented:
Could/Should we also have the following as a convenience for assigning regions of a (non-byte) array?

```
// Syntactic sugar for:
// __CPROVER_obj_upto(<expr>, s * sizeof(<type>));
__CPROVER_assignable_t __CPROVER_array_target(<type *> <expr>, size_t s);
```

Commit:
CONTRACTS: Front-end extensions for assigns clauses and frees clauses.

Add new functions to specify assignable
targets in assigns clauses:
- add type __CPROVER_assignable_t
- add builtin __CPROVER_assignable
- add builtin __CPROVER_whole_object
- add builtin __CPROVER_object_upto
- add builtin __CPROVER_typed_target
- allow function call expressions as assigns
  clause targets as long as they have the
  return type __CPROVER_assignable_t and
  are one of the built-ins.

Add support for `__CPROVER_frees()` clauses
to the contract language to let users specify
the pointers a function (or loop) is allowed
to free.

Commit:
CONTRACTS: Front-end: frees clause improvements

Add the following to the front-end:
- add `__CPROVER_freeable_t` built-in type
- add `__CPROVER_freeable` built-in function
- allow calls to `__CPROVER_freeable_t` functions
  in frees clauses
- add `__CPROVER_is_freeable` built-in predicate
- add `__CPROVER_is_freed` built-in predicate

The predicates are not yet supported in the back-end.

Commit:
CONTRACTS: Front-end extension: add `__CPROVER_is_freshr`

`__CPROVER_is_freshr` works like
`__CPROVER_is_fresh` but  takes pointers by reference
instead of by value. This willallow using the predicate
from within user-defined functions.
Only type-checking in the front-end is implemented,
support in the back-end will come with dynamic frames.

Commit:
CONTRACTS: Add class `dfcc_utilst` for DFCC utility methods.

The `dfcc_utilst` class provides methods to manipulate
function symbols in the symbol table:
- create fresh symbols
- clone and rename function parameter lists,
- clone and rename function symbols,
- add parameters to functions symbols and goto functions
These transformations will help us implement dynamic frames.

jimgrundy commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -20,25 +20,44 @@ value(s) therein are not modified.
+ of `sizeof(expr_t)` bytes starting at `&expr` is assignable:
+```c
+__CPROVER_assignable_t __CPROVER_typed_target(expr_t expr);
+```
````
Do we need this as a primitive, or could it have been a macro that we include in a proof macro header file?
#define __CPROVER_typed_target(e) __CPROVER_object_from(&e, sizeof(e))
?

jimgrundy commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -20,25 +20,44 @@ value(s) therein are not modified.
+```
+
+Given a pointer `ptr` pointing into some object `o`, 
+`__CPROVER_whole_object(ptr)` specifies that all bytes of the object `o`
````
So, for consistency and ease of memory, let's instead call this one 
__CPROVER_object_whole
That way we have 
__CPROVER_object_whole
__CPROVER_object_from
__CPROVER_object_uptop

nwetzler replied:
+1

jimgrundy commented on src/ansi-c/c_expr.h:
```diff
@@ -364,12 +364,22 @@ class function_pointer_obeys_contract_exprt : public exprt
   }
 
-  const exprt &contract() const
+  const symbol_exprt &contract_symbol_expr() const
```
Is this ok? Should we not have a test that covers this line?

jimgrundy commented on src/ansi-c/c_expr.h:
```diff
@@ -364,12 +364,22 @@ class function_pointer_obeys_contract_exprt : public exprt
-  const exprt &contract() const
+  const symbol_exprt &contract_symbol_expr() const
+  {
+    return to_symbol_expr(op1().operands().at(0));
```
Is this OK, should we not have a test that covers this line?

jimgrundy commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -777,6 +778,49 @@ void c_typecheck_baset::typecheck_declaration(
+
+        error().source_location = expr.source_location();
+        error() << id2string(id) + " is not allowed in preconditions." << eom;
+        throw 0;
```
Seems like we should have a test for this and It would be nice to see that the source location is being reported properly in the output now that we are capturing source information for contract code.

jimgrundy commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -846,6 +892,15 @@ void c_typecheck_baset::typecheck_declaration(
+        {
+          lambda_exprt lambda{temporary_parameter_symbols, frees};
+          lambda.add_source_location() = frees.source_location();
+          frees.swap(lambda);
```
Test please.

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  __CPROVER_contracts_car_t *elems;
+} __CPROVER_contracts_car_set_t;
+
+typedef __CPROVER_contracts_car_set_t *__CPROVER_contracts_car_set_ptr_t;
```
Is this type really worth defining?

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  void **elems;
+} __CPROVER_contracts_obj_set_t;
+
+typedef __CPROVER_contracts_obj_set_t *__CPROVER_contracts_obj_set_ptr_t;
```
Is this type worth defining.

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  __CPROVER_bool assert_ensures_ctx;
+} __CPROVER_contracts_write_set_t;
+
+typedef __CPROVER_contracts_write_set_t *__CPROVER_contracts_write_set_ptr_t;
```
Is this type worth defining?

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  void *ptr,
+  __CPROVER_size_t size)
+{
+  __CPROVER_assert((set != 0) & (idx < set->max_elems), "no OOB access");
```
I think I prefer "set != NULL" to "set != 0" to emphasize that set is a pointer.

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  // true iff this write set checks ensures clauses in an assumption ctx
+  __CPROVER_bool assume_ensures_ctx;
+  // true iff this write set checks ensures clauses in an assertion ctx
+  __CPROVER_bool assert_ensures_ctx;
```
Are all combinations of these booleans valid? I there a different encoding (an enum perhaps) that would naturally disallow illegal combinations?

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  __CPROVER_size_t idx = set->max_elems;
+  __CPROVER_contracts_car_t *elem = set->elems;
+CAR_SET_REMOVE_LOOP:
+  while(idx != 0)
```
I expect this will be a performance critical loop. We may want to benchmark a few different ways to write this.

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+CAR_SET_CONTAINS_LOOP:
+  while(idx != 0)
+  {
+    incl |= candidate.is_writable & elem->is_writable &
```
_might_ be faster to add
const void* elem_lb = elem->lb;
since you use that twice (would have to try though).

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  __CPROVER_size_t object_bits = __builtin_clzll(__CPROVER_max_malloc_size);
+  __CPROVER_size_t nof_objects = 1UL << object_bits;
+  set->max_elems = nof_objects;
+  set->watermark = 0;
```
I think I prefer "true" and "false" to "1" and "0" to emphasize we're talking about booleans.

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+#ifndef __CPROVER_contracts_write_set_t_defined
+#  define __CPROVER_contracts_write_set_t_defined
+
+typedef struct __CPROVER_contracts_car_t
```
Surely there is a better way to do this than repeating this code (are we repeating it? I can't quite tell). Maybe put it in a header and #include it in each place? It just seems not great.

nwetzler replied:
Is it even necessary to repeat?  Perhaps even generating this code or separating it out into different files.  This is going to be super hard to maintain.

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  __CPROVER_assert(
+    __CPROVER_POINTER_OBJECT(ptr) < set->max_elems, "no OOB access");
+  __CPROVER_size_t object_id = __CPROVER_POINTER_OBJECT(ptr);
+  set->nof_elems = set->elems[object_id] ? set->nof_elems - 1 : set->nof_elems;
```
I'd try the following to see if analysis was faster (have to try though):

set->nof_elements = set->nof_elems - (set->elems[object_id] != NULL);

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  // number of elements currently in the elems array
+  __CPROVER_size_t nof_elems;
+  // true iff nof_elems is 0
+  __CPROVER_bool is_empty;
```
Why store this if it is as simple as (nof_elems == 0)? It just seems like there is an invariant there that we have to be careful to preserve now.

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  __CPROVER_bool indexed_by_object_id;
+  // array of void *pointers, indexed by their object ID
+  // or some other order
+  void **elems;
```
This is kind of a weird structure when "indexed_by_object_id" is true, right? Each element should be either NULL or a pointer who's object bits are the same as the index and who's lower bits are 0.  We could probably do this with an array of booleans instead. Right? (But then you would have a different shape for when "indexed_by_object_id" is false so maybe this is better?)

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+  __CPROVER_assert(set->indexed_by_object_id, "indexed by object id");
+  __CPROVER_assert(
+    __CPROVER_POINTER_OBJECT(ptr) < set->max_elems, "no OOB access");
+  return set->elems[__CPROVER_POINTER_OBJECT(ptr)] == ptr;
```
Does this add up to just checking that ptr is in the set and it's offset bits are 0?

jimgrundy commented on src/ansi-c/library/cprover_contracts.c:
```diff
@@ -0,0 +1,4116 @@
+SET_CHECK_ASSIGNS_CLAUSE_INCLUSION_LOOP:
+  while(idx != 0)
+  {
+    if(current->is_writable)
```
Seems like it would be faster if you use a conditional expression or (current->is_writable) && ...

jimgrundy reviewed (commented):
That's all for now. Up to line 3000 of cprover_contracts.c

Commit:
CONTRACTS: Add library functions for dynamic frames

`cprover_contracts.c` defines C types representing
dynamic frames:
- `__CPROVER_contract_car_t`
- `__CPROVER_contract_car_set_t`
- `__CPROVER_contract_obj_set_t`
- `__CPROVER_contract_write_set_t`

It also defines functions that build and check
frame conditions in various situations:
 - allocations/deallocations,
 - assignments,
 - array operations,
 - havoc operations,
 - assigns clause inclusion checks for
   contract replacement,
 - frees clause inclusion checks for
   contract replacement,
 - etc.

It also provides C implementations for built-in predicates:
- `__CPROVER_contracts_is_freeable` for `__CPROVER_is_freeable`
- `__CPROVER_contracts_is_freed`  for `__CPROVER_is_freed`
- `__CPROVER_contracts_is_freshr` for `__CPROVER_is_fresh` and
  `__CPROVER_is_freshr`

The class `dfcc_libraryt` provides a programmatic
interface to load and use the library types and
functions and perform some optimisations like
inlining or unwinding loops in library functions.

dfcc_libraryt fixes

Commit:
CONTRACTS: Add `dfcc_contract_modet` and `dfcc_contract_handlert`.

- `dfcc_contract_modet` is an enum representing
  the CHECK and REPLACE modes for contracts,
- `dfcc_contract_handlert` is an abstract interface
  for generating GOTO instructions from contracts
  for CHECK or REPLACE modes.

Commit:
CONTRACTS: Map built-in predicate calls to their contract implem.

Rewrite calls to
- `is_fresh` to `__CPROVER_contracts_is_freshr`
- `is_freshr` to `__CPROVER_contracts_is_freshr`
- `is_freeable` to `__CPROVER_contracts_is_freeable`
- `is_freed` to `__CPROVER_contracts_is_freed`

Commit:
CONTRACTS: Add class `goto_program_cfg_infot`.

This class extends `cfg_infot` and computes locals
and dirty locals directly from a goto program's
instruction sequence. Contary to `function_cfg_infot`
it does not consider function parameters as locals.

Commit:
CONTRACTS: Add class `dfcc_instrumentt` to instrument goto
programs with checks against a dynamic frame.

Commit:
CONTRACTS: Adds the `dfcc_spec_functionst` class.

This class rewrites `__CPROVER_assignable_t`
and `__CPROVER_freeable_t` functions into
functions that accept two extra write set parameters.
Targets specified by the function are added to the
first write set.
The second write set is used to check that any other
the side effect performed by the function is allowed.
This second write set will usually be the empty write set,
or a write set that allows only writing to history variables.

Commit:
CONTRACTS: Add the `dfcc_dsl_contract_functionst` class.

The class translates the assigns and frees clauses of
a contract expressed in DSL style into goto functions
returning `__CPROVER_assignable_t` and `__CPROVER_freeable_t`
types and describing the same set of targets as the clause.

Internally it uses the `dfcc_spec_functiontst` class
to translate the functions generated from the clauses
into functions that dynamically build write sets at
runtime.

A havoc function is also generated for the assigns clause,
that havocs the locations described by the clause in
a type-directed way, taking snapshots of start addresses
from a write set instance.

nwetzler commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -1383,6 +1295,9 @@ void code_contractst::enforce_contract(const irep_idt &function)
   // Place a new symbol with the mangled name into the symbol table
+  source_locationt sl;
+  sl.set_file("instrumented for code contracts");
+  sl.set_line("0");
```
This should get fixed after a rebase on top of https://github.com/diffblue/cbmc/pull/7037, but we should make sure it doesn't revert.

nwetzler commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -1412,7 +1327,7 @@ void code_contractst::enforce_contract(const irep_idt &function)
   goto_functiont &wrapper = goto_functions.function_map[original];
   wrapper.parameter_identifiers = mangled_fun->second.parameter_identifiers;
-  wrapper.body.add(goto_programt::make_end_function());
+  wrapper.body.add(goto_programt::make_end_function(sl));
```
Same here.  https://github.com/diffblue/cbmc/pull/7037

nwetzler commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -1483,9 +1398,7 @@ void code_contractst::add_contract_check(
-  source_locationt source_location = function_symbol.location;
-  // Set function in source location to original function
-  source_location.set_function(wrapper_function);
+  const auto &source_location = function_symbol.location;
```
Same here.  https://github.com/diffblue/cbmc/pull/7037

Commit:
CONTRACTS: Add the `dfcc_dsl_wrapper_programt` class.

The class generates a sequence of GOTO instruction from
a contract expressed in the DSL syntax.

These instructions are meant to be added to the
GOTO function that serves as a wrapper for the function
under verification or being replaced by the contract.

It first uses `dfcc_dsl_contract_functionst` to generate
goto functions from the contract's assigns and frees clauses.

In contract checking mode, it generates GOTO instructions
that assume preconditions, snapshots history variables,
creates a write set instance, call the function under verification
asserts the postconditions, and returns the call's return value.

In contract replacement mode, it generates GOTO
instructions encoding the nondeterministic abstraction
defined by the contract: assert preconditions,
create and havoc assigns clause targets, create
a nondet return value for the call, nondeterministically free
frees clause targets, assume postconditions, return
a value that satisfies the post conditions.

Commit:
CONTRACTS: Add the `dfcc_dsl_contract_handlert` class.

This class implements the abstract `dfcc_contract_handlert`
interfance for DSL-style contracts using the class
`dfcc_dsl_wrapper_programt`.

Commit:
CONTRACTS: Add the `dfcc_swap_and_wrapt class`.

Given a `function_id`, a `contract_id`, a contract `mode`,
and a `dfcc_contract_handlert` for the contract,
this class swaps the body of the function to a function
with a mangled name, instruments the mangled function
for dynamic frame condition checking using `dfcc_instrument`.
The body of the original function is then replaced with
instructions generated by the contract handler for the
required contract  mode.
To handle recursive functions in contract checking mode,
the body of the wrapper function is generated in such a way
the first invocation results in checking the contract against the
mangled function and recursive invocations of the wrapper
behave like the contract abstraction (modelling that the contract holds).

Commit:
CONTRACTS: Add class `dfcct`, the main class for
function contracts with dynamic frames.

The class receives the proof harness function,
the map of functions and contracts to check
and the map of functions to replace with contracts.
It instruments the proof harness,  performs
swap-and-wrap operations for each of the checked
and replaced functions. It then instruments all other
functions for DFCC.

Commit:
CONTRACTS: add the `--dfcc` switch to the goto-instrument CLI interface

The switch expects the harness function name as parameter,
and accepts any number of `--enforce-contract` and
`--replace-call-with-contract` switches.

Commit:
CONTACTS: update regression test suite for DFCC

Updated tests
- update assigns-enforce-malloc-zero
- update assigns-local-composite
- update assigns-replace-ignored-return-value
- update assigns-replace-malloc-zero
- update assigns-slice-targets
- update assigns-slice-targets
- update assigns_enforce_01
- update assigns_enforce_02
- update assigns_enforce_03
- update assigns_enforce_04
- update assigns_enforce_05
- update assigns_enforce_06
- update assigns_enforce_07

nwetzler commented on regression/contracts/frees-clause-and-predicates/test.desc:
```diff
@@ -0,0 +1,13 @@
+^SIGNAL=0$
+--
+--
+This test checks that using the -dffdfcc flag we can check contracts that use:
```
Typo here in flag?

nwetzler commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -6,39 +6,137 @@
+A function or loop contract contract may have zero or more _assigns_ clauses.
+
+The clause accepts a semicolon-separated list of _conditional target groups_.
+A _conditional target group+ consists of an optional _condition_ followed by a
```
Markup error.  + -> _

nwetzler commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -6,39 +6,137 @@
+
+The clause accepts a semicolon-separated list of _conditional target groups_.
+A _conditional target group+ consists of an optional _condition_ followed by a
+coma-separated list of _targets_.
```
Probably not a coma.  Likely a comma-separated list.

nwetzler commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -6,39 +6,137 @@
-_assigns_ clause must not be written within the given function, even if the
-value(s) therein are not modified.
+The set of locations writable by a function is the union of the locations
+specified by the assigns clauses, or the empty set of no _assigns_ clause is
```
of -> if

nwetzler commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -6,39 +6,137 @@
+
+Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_upto(ptr, size)`
+specifies that the range of `size` bytes of `o` starting at `ptr` are assignable:
+The `size` value must such that the range does not exceed the object boundary,
```
must -> must be(?)

nwetzler commented on doc/cprover-manual/contracts-frees.md:
````diff
@@ -0,0 +1,156 @@
+__CPROVER_frees(*targets*)
+```
+
+A function or loop contract contract may have zero or more _frees_ clauses.
````
contract contract -> contract

nwetzler commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,156 @@
+
+The clause accepts a semicolon-separated list of _conditional target groups_.
+A _conditional target group+ consists of an optional _condition_ followed by a
+coma-separated list of _targets_.
```
coma -> comma

nwetzler commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,156 @@
+A function or loop contract contract may have zero or more _frees_ clauses.
+
+The clause accepts a semicolon-separated list of _conditional target groups_.
+A _conditional target group+ consists of an optional _condition_ followed by a
```
group+ -> group_

nwetzler commented on doc/cprover-manual/contracts-memory-predicates.md:
```diff
@@ -4,19 +4,62 @@
-To specify memory footprint we use a special function called `__CPROVER_is_fresh `. The meaning of `__CPROVER_is_fresh` is that we are borrowing a pointer from the
-external environment (in a precondition), or returning it to the calling context (in a postcondition).
+This allows `__CPROVER_is_freshr` to be called from within functions and still
+check or update the desired pointer:
```
I still don't think I understand the behavior or need for `__CPROVER_is_freshr`.  Can you explain why the reference version is necessary?

Also, I'm not sure the name is clear in this context.  I would expect the letters "r" and "w" to be read or write based on our other predicates, but here "freshr" means "fresh by reference".

nwetzler commented on src/goto-instrument/contracts/inlining_decorator.cpp:
```diff
@@ -0,0 +1,142 @@
+{
+  std::smatch sm;
+  std::regex_match(message, sm, no_body_regex);
+  if(sm.size() == 2)
```
Seems like there should be a better way to check this.

nwetzler commented on src/goto-instrument/contracts/dynamic-frames/dfcc_is_freeable.h:
```diff
@@ -0,0 +1,58 @@
+  /// given write_set expression as parameter to the library function.
+  void rewrite_calls(goto_programt &program, const exprt &write_set);
+
+  /// Rewrites calls to is_fresh and is_freshr predicates into calls
```
Shouldn't this be "`is_freeable` and `is_freed`"?

nwetzler commented on src/goto-instrument/contracts/dynamic-frames/dfcc_dsl_contract_functions.cpp:
```diff
@@ -0,0 +1,579 @@
+  // return type for the cloned function
+  optionalt<typet> new_return_type(library.dfcc_type[dfcc_typet::ASSIGNABLE]);
+
+  const auto &spec_function_symbol = utils.clone_and_rename_function(
```
Is there a reason to use `auto` so many times in this file?

nwetzler commented on src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.cpp:
```diff
@@ -0,0 +1,423 @@
+      if(hook_opt.has_value())
+      {
+        // TODO craft location for each instruction
+        source_locationt location;
```
What source location do we end up using here?  I don't see it assigned.

nwetzler commented on src/goto-instrument/contracts/dynamic-frames/dfcc_dsl_wrapper_program.cpp:
```diff
@@ -0,0 +1,1068 @@
+  arguments.emplace_back((exprt)false_exprt());
+
+  // assume_requires_ctx
+  arguments.emplace_back((exprt)false_exprt());
```
I'm not sure I understand why this is repeated.  Are we trying to match certain fields by installing these in order?  Is there a better way to do this?

nwetzler reviewed (commented):
Overall this code looks great and I'm excited to try it out.  I'll attempt to use it on a project and leave another comment with any issues encountered.

One concern is that it is hard to understand how all of this works together.  I feel like I'm traveling without a map.  I noticed that some new PRs have been filed that will shorten this PR once rebased, but I'd still be interested in more discussion about how this all fits together.  There are some places where the comments are nice enough (the beginning of dfcc_instrument.cpp is one example), but overall it's hard to parse what is going on without more documentation.  I'm not sure I could support this code without the original author's help.

Commit:
developer doc for contracts with dynamic frames

jimgrundy commented on regression/contracts/assigns-enforce-malloc-zero/main.c:
```diff
@@ -6,7 +6,7 @@ int foo(char *a, int size)
   __CPROVER_requires(0 <= size && size <= __CPROVER_max_malloc_size)
   __CPROVER_requires(a == NULL || __CPROVER_is_fresh(a, size))
-  __CPROVER_assigns(a: __CPROVER_POINTER_OBJECT(a))
+  __CPROVER_assigns(a: __CPROVER_whole_object(a))
```
I would really like this to be named "__CPROVER_object_whole" for consistency with "__CPROVER_object_from" and "__CPROVER_object_slice"

jimgrundy commented on regression/contracts/frees-clause-and-predicates-fail/main.c:
```diff
@@ -0,0 +1,46 @@
+#include <stdlib.h>
+
+// A function defining a conditionally freeable target
+__CPROVER_freeable_t
```
I thought you were moving away from using these "types"?

jimgrundy commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -216,6 +216,29 @@ void ansi_c_internal_additions(std::string &code)
+    // Declares bytes from ptr to the end of the object as assignable
+    CPROVER_PREFIX "assignable_t " CPROVER_PREFIX "object_from(void *ptr);\n"
+    // Declares the whole object pointer to by ptr
+    CPROVER_PREFIX "assignable_t " CPROVER_PREFIX "whole_object(void *ptr);\n"
```
whole_object -> object_whole

jimgrundy commented on src/goto-instrument/contracts/doc/user-guide/contracts-assigns.md:
```diff
@@ -6,39 +6,137 @@
+The clause accepts a semicolon-separated list of _conditional target groups_.
+A _conditional target group+ consists of an optional _condition_ followed by a
+coma-separated list of _targets_.
+A _target_ is either an lvalue expression or a call to a function returning
```
Are we keeping that?

jimgrundy commented on src/goto-instrument/contracts/doc/user-guide/contracts-assigns.md:
```diff
@@ -6,39 +6,137 @@
-_assigns_ clause must not be written within the given function, even if the
-value(s) therein are not modified.
+The set of locations writable by a function is the union of the locations
+specified by the assigns clauses, or the empty set of no _assigns_ clause is
```
"of" -> "if"

jimgrundy commented on src/goto-instrument/contracts/doc/user-guide/contracts-assigns.md:
```diff
@@ -6,39 +6,137 @@
+
+Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_upto(ptr, size)`
+specifies that the range of `size` bytes of `o` starting at `ptr` are assignable:
+The `size` value must such that the range does not exceed the object boundary,
```
"must such" -> "must be such"

jimgrundy commented on src/goto-instrument/contracts/doc/user-guide/contracts-assigns.md:
````diff
@@ -6,39 +6,137 @@
+__CPROVER_assignable_t __CPROVER_object_from(void *ptr);
+```
+
+Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_upto(ptr, size)`
````
Can we go back to the name "object_slice" instead of "object_upto".
Or, can we rename "havoc_slice" to "havoc_upto"? These things should have related names.

jimgrundy commented on src/goto-instrument/contracts/doc/user-guide/contracts-assigns.md:
```diff
@@ -6,39 +6,137 @@
+call-by-contract replacement, `__CPROVER_havoc_slice(ptr, size)` is used to
+havoc these byte ranges, and `__CPROVER_havoc_slice` does not support
+havocing pointers. `__CPROVER_typed_target` must be used to specify targets
+that are pointers.
```
Can we have a version of typed_target that also takes a size. Maybe call it "typed_array"? I have seen examples where we have an array of pointers and it would be nice to have a convenient way to deal with those.

jimgrundy commented on src/goto-instrument/contracts/doc/user-guide/contracts-frees.md:
````diff
@@ -0,0 +1,156 @@
+The predicate:
+
+```c
+__CPROVER_bool __CPROVER_is_freeable(void *ptr);
````
Does it also make sense to have a version of this that takes a size for describing code that works with arrays of pointers? (or do we handle that with a loop?)

jimgrundy replied:
But if it took a size could we handle a symbolic number of frees?

jimgrundy commented on src/goto-instrument/contracts/doc/user-guide/contracts-frees.md:
````diff
@@ -0,0 +1,156 @@
+The predicate:
+
+```c
+__CPROVER_bool __CPROVER_is_freed(void *ptr);
````
It breaks my intuitions that we would have a construct that worked in pre-conditions and not post-conditions or visa-versa as I expect these to be duals and to place the opposite roles when checking or using contracts, to if it can appear in one role in one context it must work in the other role in the opposite context, no?

jimgrundy commented on src/goto-instrument/contracts/dynamic-frames/dfcc.cpp:
```diff
@@ -0,0 +1,479 @@
+
+\*******************************************************************/
+
+// TODO havoc global statics
```
This should be done before merging.

jimgrundy reviewed (commented):
Are we planning to support two version of contract use/checking going forward? With and without --dfcc? If so, why?

Commit:
dfcc_swap_and_wrap fixes

Commit:
dfcc_dsl_contract_functionst fixes

feliperodri commented:
Fixes https://github.com/diffblue/cbmc/issues/7197

feliperodri commented:
Fixes https://github.com/diffblue/cbmc/issues/7198

feliperodri commented:
Fixes https://github.com/diffblue/cbmc/issues/7206

remi-delmas-3000 commented:
Superseded by #7242
